File: Medium/2812. Find the Safest Path in a Grid.java
class Solution {
    private final static int[] dirs = { 1, 0, -1, 0, 1 };

    public int maximumSafenessFactor(List<List<Integer>> grid) {
        final int N = grid.size();
        final var matrix = new int[N][N];
        final var queue = new ArrayDeque<int[]>();
        final var scores = new int[N][N];
        for (int r = 0; r < N; r++) {
            for (int c = 0; c < N; c++) {
                if (grid.get(r).get(c) == 1) {
                    matrix[r][c] = 1;
                    queue.offer(new int[] { r, c });
                    scores[r][c] = 0;
                } else {
                    scores[r][c] = Integer.MAX_VALUE;
                }
            }
        }
        while (!queue.isEmpty()) {
            var curr = queue.poll();
            int cr = curr[0], cc = curr[1];
            for (int i = 0; i < 4; i++) {
                int nr = cr + dirs[i], nc = cc + dirs[i + 1];
                if (nr < 0 || nc < 0 || nr >= N || nc >= N || scores[nr][nc] <= scores[cr][cc] + 1) {
                    continue;
                }
                scores[nr][nc] = scores[cr][cc] + 1;
                queue.offer(new int[] { nr, nc });
            }
        }

        final var pq = new PriorityQueue<int[]>((x, y) -> Integer.compare(y[0], x[0]));
        pq.offer(new int[] { scores[0][0], 0, 0 });
        matrix[0][0] = 2;
        while (!pq.isEmpty()) {
            var curr = pq.poll();
            int cs = curr[0], cr = curr[1], cc = curr[2];
            if (cr == N - 1 && cc == N - 1) {
                return cs;
            }
            for (int i = 0; i < 4; i++) {
                int nr = cr + dirs[i], nc = cc + dirs[i + 1];
                if (nr < 0 || nc < 0 || nr >= N || nc >= N || matrix[nr][nc] == 2) {
                    continue;
                }
                pq.offer(new int[] { Math.min(cs, scores[nr][nc]), nr, nc });
                matrix[nr][nc] = 2;
            }
        }
        return -1;
    }
}

File: Medium/2812. Find the Safest Path in a Grid.py
DIRS = [[0, 1], [0, -1], [1, 0], [-1, 0]]

class Solution:
    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
        if grid[0][0] or grid[-1][-1]:
            return 0
        N, Q = len(grid), deque()
        for r in range(N):
            for c in range(N):
                if grid[r][c]:
                    Q.append((r, c))
        while Q:
            r, c = Q.popleft()
            safety = grid[r][c]
            for dr, dc in DIRS:
                nr, nc = r + dr, c + dc
                if 0 <= nr < N and 0 <= nc < N and not grid[nr][nc]:
                    grid[nr][nc] = safety + 1
                    Q.append((nr, nc))
        PQ = [(-grid[0][0], 0, 0)]
        while PQ:
            safety, r, c = heapq.heappop(PQ)
            safety *= -1
            if r == N - 1 and c == N - 1:
                return safety - 1
            for dr, dc in DIRS:
                nr, nc = r + dr, c + dc
                if 0 <= nr < N and 0 <= nc < N and grid[nr][nc] > 0:
                    heapq.heappush(PQ, (-min(safety, grid[nr][nc]), nr, nc))
                    grid[nr][nc] *= -1
        return grid[-1][-1] - 1
